### ISSUE TITLE
Shoud the nc file add CRS according to cf-conventions? 

### ISSUE THREAD

**Issue by singledoggy** [miniufo/xgrads#36]
Recently I've learned the [cf-conventions](https://pyproj4.github.io/pyproj/stable/api/crs.html#pyproj.crs.CRS.from_cf) that defines a way to store crs in  `.nc` file and  [pyproj](https://pyproj4.github.io/pyproj/latest/build_crs_cf.html) gives a method to get crs from and to nc. 
So is it possible to add crs when converting `.nc`?  Otherwise we still need the original `.ctl ` files to store the projection information. It's weird. 
Thanks again for your work. It helps a lot in my work.

**Comment by miniufo** [comment]
That is interesting.  Could you provide a more concrete example on how to do this?  I just go through the doc you send, and the simplest way is to convert the crs in `.ctl` to a string of comments that will be put into the `.nc`?  Then it is easy to convert it back using `pyproj`.  Am I right?

**Comment by singledoggy** [comment]
I'm not a specialist of that too, so I guess you're right. And I don't know the details of this conventions and if it will become popular soon. As I know, many `.nc`files still don't save the projection information now.  I just want to save compelete informations in `xr.DataArray` or the `.nc` files. 

Xarray mentioned [CF](https://xarray.pydata.org/en/v2022.03.0/user-guide/weather-climate.html?highlight=CF%20projection#related-variables) data just here, and there are few informations. 


But this snippet returens `True` so at least this string can be safely convert back?
```
from pyproj import CRS
from xgrads import get_data_projection
crs = get_data_projection(ctl)
cf = crs.to_cf()
crs_new = crs.from_cf(cf)
crs_new == crs
```
It seems to use an attribute named` grid_mapping` to store this.

**Comment by miniufo** [comment]
>crs = get_data_projection(ctl)
>cf = crs.to_cf()

`get_data_projection` does not return a CRS but a `<cartopy.crs.LambertConformal at 0x11920270>`.  So I cannot find `to_cf()`.  Have you tried this snippet?

**Comment by singledoggy** [comment]
I test that code in an environment contains ` cartopy==0.20.1 `, but in another version ` cartopy==0.17 ` it  returns the same as your comment.
<img width="694" alt="image" src="https://user-images.githubusercontent.com/44222677/160312098-8cab7470-934b-4ad0-b4f1-9564a69d80b9.png">
It seems that [cartopy 0.20+ is required](https://pyproj4.github.io/pyproj/latest/crs_compatibility.html#cartopy), and 
> cartopy.crs.CRS inherits from pyproj.crs.CRS, so it should behave like a pyproj.crs.CRS.

It may be a pretty new feature? But I know it sometimes cause compatibility issues.

**Comment by miniufo** [comment]
I guess this is really new feature.  The doc shows that api's name is `cs_to_cf()` at [here](https://scitools.org.uk/cartopy/docs/latest//reference/generated/cartopy.crs.CRS.html).

I currently cannot make any test because I don't have the newest environment.  From the codes you showed it is pretty simple to do this.  If `cf` is a string, adding it to the dataset before writting it to a disk file would be OK.

It seems that everything is changing to take into account the CF convention, although I knew it many years ago.

**Comment by miniufo** [comment]
When I tried to solve #37, I found a workaround for this.  I use cartopy 0.18.0 and the returned crs from `get_data_projection` is not a subclass of `pyproj.crs.CRS`.  But we can create one by:
```python
from pyproj import CRS

crs = get_data_projection(ctl, Rearth=6370000)
proj_crs = CRS(crs.proj4_init)
cf = proj_crs.to_cf()
```

I get `cf` as:
![image](https://user-images.githubusercontent.com/9312831/167233161-ab61cf47-5818-4da2-9984-6ab1be584bf5.png)

It is a little too long and not a string.  But `crs.proj4_init` is a short string as:
![image](https://user-images.githubusercontent.com/9312831/167233194-165b47b6-2721-4181-96d6-61aa61c92d6a.png)

So if you want to put cf into nc file, you can put `crs.proj4_init` in a `grid_mapping` comment, and recover a CRS from the above code.  I guess that cartopy is doing a significant update (I re-install anaconda and all python package and still get cartopy 0.18.0).  We shall wait to see if the API's name is eventually fixed (e.g., the link you provided on 24 Mar at the very beginning is now broken...).

**Comment by singledoggy** [comment]
Yes, I've found it's an unsolved issue of Cartopy for a long time. And [Xarray don't read the crs data by default](https://github.com/pydata/xarray/issues/2288). I've tried [rioxarray](https://corteva.github.io/rioxarray/latest/getting_started/crs_management.html) to get the crs, but there are still some bugs.
So maybe this is their work to get a API to convert it.

**Comment by miniufo** [comment]
Hi @Mo-Dabao, that's very useful.  Thank you for sharing this point.  Recently, I've added a function to parse the global comments in ctl files output by WRF (see (here)[https://github.com/miniufo/xgrads/blob/ed26d5ed93d18b417b59320c17b1ab0568fa8c2c/xgrads/core.py#L437]).

I may add a util function in `utils.py` as (pseudo codes):
```python
def to_CFnetcdf(dset, comments, Rearth, path):
    # add_comments_to_dset(dset, comments)
    # add_Rearth_to_dset(dset, Rearth)
    # dset.to_netcdf(path)
```

Do you think this is OK? Can you please try if the output NC file can be recognized by panoply, as I don't have panoply?

**Comment by Mo-Dabao** [comment]
> I may add a util function in `utils.py` as (pseudo codes):
> 
> ```python
> def to_CFnetcdf(dset, comments, Rearth, path):
>     # add_comments_to_dset(dset, comments)
>     # add_Rearth_to_dset(dset, Rearth)
>     # dset.to_netcdf(path)
> ```
> 
> Do you think this is OK?

It's totally up to you. But the `earth_radius` I mentioned before may only apply to those data based on "spherical" Earth, for example some WRF outputs. So it might take more design to be compatible to all kinds of Earth.

> Can you please try if the output NC file can be recognized by panoply, as I don't have panoply?

`panoply` can be download free from NASA's web site [here](https://www.giss.nasa.gov/tools/panoply/). And it runs without installing, as long as you have JAVA 9 installed first.
